### ISSUE TITLE
Microsoft.Extensions.Http is too heavy-weight for mobile usages

### ISSUE THREAD

**Issue by eerhardt** [dotnet/runtime#66863]
### Description

In profiling the [.NET Podcasts application](https://github.com/microsoft/dotnet-podcasts) startup time on an Android Pixel 5a, I'm seeing its [usage](https://github.com/microsoft/dotnet-podcasts/blob/392453be5fa55c47ff84b98731d9b71fa6dc489c/src/Mobile/Services/ServicesExtensions.cs#L11-L14) of a named HttpClient using ` builder.Services.AddHttpClient<ShowsService>(client => {})` as taking up 5% of the startup cost of the app.

### Configuration

* Which version of .NET is the code running on? `6.0`
* What OS version, and what distro if applicable? `Android 12`
* What is the architecture (x64, x86, ARM, ARM64)? `ARM64`
* If relevant, what are the specs of the machine? `Google Pixel 5a`

### Regression?

No

### Data

Using the startup profiling script at https://github.com/jonathanpeppers/maui-profiling#faq, I'm getting the following results:

(using a full AOT of the app, so to eliminate any JIT time)

#### Startup time with AddHttpClient

Average(ms): 838.2
Average(ms): 832.4
Average(ms): 835.2

#### Startup time with just creating a shared HttpClient without DI

Average(ms): 795.5
Average(ms): 799.8
Average(ms): 797.9

### Analysis

[dotnet-podcasts.speedscope.zip](https://github.com/dotnet/runtime/files/8307769/dotnet-podcasts.speedscope.zip)

1. Looking through the attached speedscope, a large chunk of time is spent creating the named Http client service in [AddTransientHelper](https://github.com/dotnet/runtime/blob/0b12d37843e7165fb4c8b794186f19ef43af6c73/src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs#L260):

![image](https://user-images.githubusercontent.com/8291187/159086734-4f7f70cd-ef20-404b-b077-ec8f4d2e04e5.png)

A big chunk of that time is due to calling [ActivatorUtilities.CreateFactory](https://github.com/dotnet/runtime/blob/0b12d37843e7165fb4c8b794186f19ef43af6c73/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs#L91), which uses System.Linq.Expressions and IL Ref.Emit.

2. The Logging in Microsoft.Extensions.Http also seems to be adding a decent amount of overhead:

![image](https://user-images.githubusercontent.com/8291187/159087742-37a4297e-4b14-4f5b-bb21-318acaf77602.png)

### Discussion

I wonder if using Microsoft.Extensions.Http on a mobile application is recommended at all. Or if a mobile app should just be following the guidance we give in [our docs](https://docs.microsoft.com/dotnet/api/system.net.http.httpclient):

> HttpClient is intended to be instantiated once and re-used throughout the life of an application. 

@davidfowl @jonathanpeppers @dotnet/ncl

**Comment by jonathanpeppers** [comment]
For the podcast app, should we maybe go with something like:
```
services.TryAddSingleton<HttpClient>(sp => new HttpClient());
```
And use 1 instance of `HttpClient` for the entire app?

**Comment by CarnaViire** [comment]
Yes, I would say it's quite a heavy overhead for the mobile usages - which might not be needed at all.
The factory means to solve two problems - socket exhaustion (that can be solved by a static/singleton HttpClient) and not losing DNS changes. I am not sure whether the second problem even exists for mobile usages? Given it's not SocketsHttpHandler that is used there.

**Comment by jonathanpeppers** [comment]
Socket exhaustion definitely seems like something server scenarios would run into. If mobile app is making enough web requests to hit that problem, I would probably recommend "don't do that". Even if a mobile app was downloading lots of files, they should probably limit to 4 (or some number) of concurrent connections

**Comment by eerhardt** [comment]
> It just so happens that startup time is affected because of IL emit

The Logging in Microsoft.Extensions.Http also seems to be adding a decent amount of overhead

**Comment by davidfowl** [comment]
Hmmm, maybe worth making that opt-in.
